### PR TITLE
feat(graphql): support http get and post json request

### LINKED ISSUE #6266: bug: ctx.lua#59 parse_graphql(ctx)
### Issue description

use whole request body to parse graphql will get parse error. 
graphql request body is json , example :`{"query":"query{getUser{name age}}","variables":null}`, not `query{getUser{name age}}`

### Environment

- apisix version (cmd: `apisix version`):  `
apache/apisix:2.12.0-alpine`
- OS (cmd: `uname -a`): `docker `
- OpenResty / Nginx version (cmd: `nginx -V` or `openresty -V`):  null
- etcd version, if have (cmd: run `curl http://127.0.0.1:9090/v1/server_info` to get the info from server-info API):  `bitnami/etcd:3.4.15`
- apisix-dashboard version, if have: `
apache/apisix-dashboard:2.10.1-alpine`
- the plugin runner version, if the issue is about a plugin runner (cmd: depended on the kind of runner):
- luarocks version, if the issue is about installation (cmd: `luarocks --version`):

### Steps to reproduce

1. define graphql
```graphql
query {
    getUser:User
}

type User{
    name:String
    age:String
}
```
2. add route
```bash
curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -i -d '
{
    "methods": ["POST"],
    "uri": "/graphql",
    "vars": [
        ["graphql_operation", "==", "query"],
        ["graphql_name", "==", "getUser"]
    ],
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "127.0.0.1:1980": 1
        }
    }
}'
```
3. perform graphql request by curl
```bash
curl 'http://127.0.0.1:9080/graphql' \
  -H 'Content-Type: text/plain;charset=UTF-8' \
  -H 'Accept: */*' \
  --data-raw '{"query":"query getUser{getUser{name age}}","variables":null}' \
  --compressed
```

### Actual result

HTTP/1.1 404 Not Found
Date: Tue, 08 Feb 2022 07:39:16 GMT
Content-Type: text/plain; charset=utf-8
Transfer-Encoding: chunked
Connection: keep-alive

{
  "error_msg": "404 Route Not Found"
}

### Error log

2022/02/08 07:39:16 [error] 45#45: *1085159 [lua] ctx.lua:80: get_parsed_graphql(): failed to parse graphql: Syntax error near line 1 body: 

### Expected result

success

### PR DESCRIPTION
**Pull request by shuaijinchao**
### What this PR does / why we need it:
FIX #6266

### Pre-submission checklist:
* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: apisix/core/ctx.lua [hunk 1]
```diff
@@ -36,7 +37,48 @@ local pcall        = pcall
...
+        return body
+    end,
+
+    [GRAPHQL_REQ_METHOD_HTTP_POST] = function(ctx, max_size)
+        local body, err = request.get_body(max_size, ctx)
+        if not body then
+            return nil, "failed to read graphql data, " .. (err or "request body has zero size")
+        end
+
+        if request.header(nil, "Content-Type") == GRAPHQL_REQ_MIME_JSON then
```
Comment (spacewander) [inline apisix/core/ctx.lua:64]: We can pass ctx here?

File: apisix/core/ctx.lua [hunk 2]
```diff
@@ -36,7 +37,48 @@ local pcall        = pcall
...
+local GRAPHQL_DEFAULT_MAX_SIZE       = 1048576               -- 1MiB
+local GRAPHQL_REQ_DATA_KEY           = "query"
+local GRAPHQL_REQ_METHOD_HTTP_GET    = "GET"
+local GRAPHQL_REQ_METHOD_HTTP_POST   = "POST"
+local GRAPHQL_REQ_MIME_JSON          = "application/json"
+
+
+local fetch_graphql_data = {
+    [GRAPHQL_REQ_METHOD_HTTP_GET] = function(ctx, max_size)
+        local body = request.get_uri_args(ctx)[GRAPHQL_REQ_DATA_KEY]
```
Comment (spacewander) [inline apisix/core/ctx.lua:49]: The body here might be a table

File: apisix/core/ctx.lua (hunk 1 above)
Comment (shuaijinchao) [reply apisix/core/ctx.lua:64]: Of course, I update it later.

File: apisix/core/ctx.lua (hunk 2 above)
Comment (shuaijinchao) [reply apisix/core/ctx.lua:49]: if it is a table, do you think should we use the first value in the table or reject the request.

Comment (spacewander) [reply apisix/core/ctx.lua:49]: We can use the first one

Comment (shuaijinchao) [reply apisix/core/ctx.lua:49]: updated

File: apisix/core/ctx.lua (hunk 1 above)
Comment (shuaijinchao) [reply apisix/core/ctx.lua]: updated

**Review by spacewander — APPROVED** [review]

**Review by bzp2010 — APPROVED** [review]

**Pull request merged**